table/block_based/block_based_table_reader.cc

```diff
@@ -1941,12 +1941,14 @@ bool BlockBasedTable::PrefixMayMatch(
   FilterBlockReader* const filter = rep_->filter.get();
   bool filter_checked = true;
   if (filter != nullptr) {
+    const bool no_io = read_options.read_tier == kBlockCacheTier;
+
     if (!filter->IsBlockBased()) {
       const Slice* const const_ikey_ptr = &internal_key;
       may_match = filter->RangeMayExist(
           read_options.iterate_upper_bound, user_key, prefix_extractor,
           rep_->internal_comparator.user_comparator(), const_ikey_ptr,
-          &filter_checked, need_upper_bound_check, lookup_context);
+          &filter_checked, need_upper_bound_check, no_io, lookup_context);
```

Comment:
Based on the comment at the top of this method, I think we should pass in `true` unconditionally for `no_io` here.

Reply:
Always no I/O means that once the filter block is evicted, we'll never use filters anymore? The behavior feels a little bit odd to me. Also, I worry about changing behavior in a new release even if the behavior is a "bug". Now sure @pdillinger has any thoughts on it.

Reply:
Note that the other branch explicitly sets `kBlockCacheTier`. BTW, this branch is for the deprecated block based filter.

Reply:
The comment is "Otherwise, this method guarantees no I/O will be incurred."? I interpreted the "otherwise" to mean "if the above properties don't hold", but I guess that's not right.

Why does this method need to avoid IO?

What Siying said sounds like a big concern: we use `cache_index_and_filter_blocks = true`, large `max_open_files`, almost no compactions, and we churn through block cache pretty quickly. So sst files probably stay open for a long time and the prefetched bloom filters probably get evicted pretty quickly. We do use `cache_index_and_filter_blocks_with_high_priority` (or rather a hack on our side that does almost the same thing), but the high-priority part of the cache probably gets plenty of evictions too.

Reply:
The `kBlockCacheTier` in the other branch is not as bad because it only needs the *index* block to be in cache, and index block is pulled into cache by most reads, so if you call Get() or Seek() in this CF often enough the index block will be in cache most of the time.

Reply:
My guess is that the requirement that PrefixMayMatch() doesn't do IO is left over from before `cache_index_and_filter_blocks` existed, and is now obsolete, and harmful when `cache_index_and_filter_blocks = true`.

I looked through the relevant code and couldn't find anything that would break from doing IO here. BlockBasedTable::PrefixMayMatch() is called only by BlockBasedTableIterator seeks (which are ok with IO). PrefixMayMatch() calls FilterBlockReader::PrefixMayMatch or FilterBlockReader::RangeMayExist. I checked all implementations of these two methods (BlockBasedFilterBlockReader, PartitionedFilterBlockReader, FullFilterBlockReader) and didn't find anything that would break with `no_io = false`.

The main concern is reading the filter block and *not* inserting it into block cache - re-reading the filter block on each iterator seek would probably be more expensive than disabling filter altogether. As far as I could tell, this won't happen for reasonable combinations of options:
  (1) cache_index_and_filter_blocks = false, non-partitioned filter: will fetch filter block when opening file and keep it outside the cache,
  (2) cache_index_and_filter_blocks = true, block_cache not nullptr: if filter block is read, it'll be inserted into block cache.
The remaining weird combinations of options will cause iterators to re-read the filter block on each seek:
  (3) cache_index_and_filter_blocks = false, partitioned filter: this combination doesn't seems useful, the whole point of partitioned filter is to not keep it all in memory,
  (4) cache_index_and_filter_blocks = true, block_cache = nullptr.

That seems acceptable to me. Updated the PR to remove the comment about PrefixMayMatch() not doing IO, making it officially do IO. Lmk if I missed something.

Reply:
Thanks for looking into this @al13n321 . For the record, in case partitioned filters/indexes are used, `cache_index_and_filter_blocks` affects only the top-level block but not the partitions; also, the combination in (4) is caught during parameter sanitization.